monero-sys/src/lib.rs

```diff
@@ -0,0 +1,1511 @@
+            let (synced, sync_progress) =
+                { (self.synchronized().await, self.sync_progress().await) };
+
+            // Notify the listener (if it exists)
+            if sync_progress > current_progress {
+                if let Some(listener) = &listener {
+                    listener(sync_progress);
+                }
+            }
+
+            // Update the current progress
+            current_progress = sync_progress;
+
+            // If the wallet is synced, break out of the loop.
+            if synced {
+                break;
+            }
+
+            tracing::trace!(
+                %sync_progress,
+                "Wallet sync not complete, sleeping for {}ms",
+                POLL_INTERVAL_MILLIS
+            );
+
+            // Otherwise, sleep for a bit and try again.
+            tokio::time::sleep(std::time::Duration::from_millis(POLL_INTERVAL_MILLIS)).await;
+        }
+
+        tracing::debug!("Wallet synced");
+
+        Ok(())
+    }
+
+    /// Check the status of a transaction.
+    async fn check_tx_status(
+        &self,
+        txid: String,
+        tx_key: monero::PrivateKey,
+        destination_address: &monero::Address,
+    ) -> anyhow::Result<TxStatus> {
+        let destination_address = *destination_address;
+        self.call(move |wallet| wallet.check_tx_status(&txid, tx_key, &destination_address))
+            .await
+    }
+
+    /// Scan a transaction for the wallet.
+    /// This makes a transaction visible to the wallet without requiring a full sync.
+    pub async fn scan_transaction(&self, txid: String) -> anyhow::Result<()> {
+        self.call(move |wallet| wallet.scan_transaction(txid)).await
+    }
+
+    /// Wait until a transaction is confirmed.
+    pub async fn wait_until_confirmed(
+        &self,
+        txid: String,
+        tx_key: monero::PrivateKey,
+        destination_address: &monero::Address,
+        expected_amount: monero::Amount,
+        confirmations: u64,
+        listener: Option<impl Fn(u64) + Send + 'static>,
+    ) -> anyhow::Result<()> {
+        const DEFAULT_CHECK_INTERVAL_SECS: u64 = 15;
+
+        let mut poll_interval = tokio::time::interval(tokio::time::Duration::from_secs(
+            DEFAULT_CHECK_INTERVAL_SECS,
+        ));
+
+        loop {
+            poll_interval.tick().await;
+
+            let tx_status = match self
+                .check_tx_status(txid.clone(), tx_key, destination_address)
+                .await
+            {
+                Ok(tx_status) => tx_status,
+                Err(e) => {
+                    tracing::error!(
+                        "Failed to check tx status: {}, rechecking in {}s",
+                        e,
+                        DEFAULT_CHECK_INTERVAL_SECS
+                    );
+                    continue;
+                }
+            };
+
+            // Make sure the amount is correct
+            if tx_status.received != expected_amount {
+                tracing::error!(
```

Comment:
We don't need to log here if the error is passed to the caller.

At most at `trace` level.